content/courses/advanced-ada/parts/resource_management/controlled_types.rst

```diff
@@ -2343,6 +2343,499 @@ is allocated in the :ada:`Initialize` procedure.
+storing a new element into the stack. If the stack is full, we call the
+:ada:`Increase_Size` procedure to *increase* the size of the array. This is
+actually done by calling the :ada:`Reallocate_Data` procedure, which allocates
+a new array for the stack and copies the original data to the new array.
+
+Also, when copying an unbounded stack object to another object of this type, a
+call to the :ada:`Adjust` procedure is triggered |mdash| we do this by the
+assignment :ada:`Stack_2 := Stack` in the :ada:`Show_Unbounded_Stack`
+procedure. In the :ada:`Adjust` procedure, we call the
+:ada:`Allocate_Duplicate_Data` procedure to allocate a new array for the stack
+data and copy the data from the original stack. (Internally, the
+:ada:`Allocate_Duplicate_Data` procedure calls the :ada:`Reallocate_Data`
+procedure, which we already mentioned.)
+
+By encapsulating the access type handling in controlled types, we can ensure
+that the access objects are handled correctly: no incorrect pointer usage or
+memory leak can happen when we use this strategy.
+
+
+Encapsulating file handling
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Controlled types can be used to encapsulate file handling, so that files are
+automatically created and closed. A common use-case is when a new file is
+expected to be created or opened when we declare the controlled object, and
+closed when the controlled object gets out of scope.
+
+A simple example is the one of a logger, which we can use to write to a
+logfile by simple calls to :ada:`Put_Line`:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Resource_Management.Controlled_Types.Applications.Logger main=show_logger.adb
+
+    with Ada.Text_IO; use Ada.Text_IO;
+    with Ada.Finalization;
+
+    package Loggers is
+
+       type Logger (<>) is
+         limited private;
+
+       function Init (Filename : String)
+                      return Logger;
+
+       procedure Put_Line (L : Logger;
+                           S : String);
+
+    private
+
+       type Logger is new
+         Ada.Finalization.Limited_Controlled with
+          record
+             Logfile : File_Type;
+          end record;
+
+       procedure Finalize
+         (L : in out Logger);
+
+    end Loggers;
+
+    package body Loggers is
+
+       --
+       --  SUBPROGRAMS
+       --
+
+       function Init (Filename : String)
+                      return Logger is
+       begin
+          return L : Logger do
+             Create (L.Logfile, Out_File, Filename);
+          end return;
+       end Init;
+
+       procedure Put_Line (L : Logger;
+                           S : String) is
+       begin
+          Put_Line ("Logger: Put_Line");
+          Put_Line (L.Logfile, S);
+       end Put_Line;
+
+       --
+       --  PRIVATE SUBPROGRAMS
+       --
+
+       procedure Finalize
+         (L : in out Logger) is
+       begin
+          Put_Line ("Finalizing Logger...");
+          Close (L.Logfile);
```

Comment:
Same issue of multiple calls possible, so a check is needed to determine whether the file is already closed (Status_Error is raised if it is not open).